### PR TITLE
I825 no additional entry points

### PR DESCRIPTION
**Pull request by Jesse-Good**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x ] Tests for the changes have been added (for bug fixes / features)
- [x ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
rollup_bundle build rule requires additional_entry_points explicitly listed.

Issue Number: 825

## What is the new behavior?

rollup_bundle adds boolean parameter auto_additional_entry_points, which can be used instead of additional_entry_points. auto_additional_entry_points is True by default.

## Does this PR introduce a breaking change?

- [ x] Yes
- [ ] No

Existing invocations of rollup_bundle() that omit additional_entry_points may need to set auto_additional_entry_points to False.

### CODE REVIEW

File: internal/e2e/rollup_code_splitting/BUILD.bazel [hunk 1]
```diff
@@ -11,18 +11,32 @@ rollup_bundle(
     license_banner = ":license.txt",
 )
 
+rollup_bundle(
```
Comment (alexeagle) [inline internal/e2e/rollup_code_splitting/BUILD.bazel]: hmm, I think we want to just change most of the tests to remove addl_entry_points and test the auto behavior everywhere, except maybe for one test that's left on the legacy behavior just to make sure we don't break it during the transition.

File: internal/web_package/test/BUILD.bazel [hunk 2]
```diff
@@ -5,6 +5,7 @@ package(default_visibility = ["//visibility:public"])
 
 rollup_bundle(
     name = "bundle",
+    auto_additional_entry_points = False,
```
Comment (alexeagle) [inline internal/web_package/test/BUILD.bazel]: in particular I think we want to have it True here - I'd like to understand if that's breaking by changing the tests to illustrate just what that breakage is

File: internal/e2e/rollup_code_splitting/BUILD.bazel (hunk 1 above)
Comment (Jesse-Good) [reply internal/e2e/rollup_code_splitting/BUILD.bazel]: Ok, here's another branch where I just convert the test to auto, deleting test coverage for (non auto) additional entry points. I have to delete the tests for additional_entry.js to get this to work. A new chunk-X.js is created (it's working) but I think I can only write a test for it if put X (the hash) into the test. Do you think that's OK? 

https://github.com/bazelbuild/rules_nodejs/compare/master...Jesse-Good:i825-autoEntryPoints2?expand=1

**Comment by Jesse-Good** [comment]
> The new test doesn't demonstrate that code splitting is happening - are there actually additional chunks produced?

Yes, chunk is being produced but with name chunk-X.js. Is it safe to write a test against the hashed X value?

File: internal/e2e/rollup_code_splitting/BUILD.bazel (hunk 1 above)
Comment (Jesse-Good) [reply internal/e2e/rollup_code_splitting/BUILD.bazel]: PTAL. I've updated this branch.

File: internal/web_package/test/BUILD.bazel (hunk 2 above)
Comment (Jesse-Good) [reply internal/web_package/test/BUILD.bazel]: //internal/web_package/test2:test fails because the <script> tags look different. There's a bunch of <script type="module" and nomodule in the actual output but the test is not expecting that. 
[test.log](https://github.com/bazelbuild/rules_nodejs/files/3298439/test.log)

Comment (alexeagle) [reply internal/web_package/test/BUILD.bazel]: paired on this in person - we now understand that the golden file changes to the HTML are desirable - enabling code splitting had the side effect of enabling differential loading

File: internal/rollup/rollup_bundle.bzl [hunk 3]
```diff
@@ -611,10 +611,19 @@ ROLLUP_ATTRS = {
         It is sufficient to load one of these SystemJS boilerplate/entry point
         files as a script in your HTML to load your application""",
     ),
+    "auto_additional_entry_points": attr.bool(
```
Comment (alexeagle) [inline internal/rollup/rollup_bundle.bzl]: decided to rename it "enable_code_splitting" since this is orthogonal to the number of entry points

File: internal/rollup/rollup_bundle.bzl [hunk 4]
```diff
@@ -611,10 +611,19 @@ ROLLUP_ATTRS = {
...
+        doc = """Set this to True instead of specifying additional_entry_points to
```
Comment (alexeagle) [inline internal/rollup/rollup_bundle.bzl]: add a link to https://rollupjs.org/guide/en#code-splitting

and "code splitting is always enabled when additional_entry_points is non-empty"

File: internal/e2e/rollup_code_splitting/BUILD.bazel (hunk 1 above)
Comment (Jesse-Good) [reply internal/e2e/rollup_code_splitting/BUILD.bazel]: Ok, I've updated :bundle to enable code splitting. And I added a new test to cover the previous behavior with specifying additional_entry_points.

**Comment by Jesse-Good** [comment]
PTAL, merged from origin and responded to comments.

File: internal/rollup/rollup.config.js [hunk 5]
```diff
@@ -206,17 +202,16 @@ const config = {
   preserveSymlinks: true,
 }
 
-if (enableCodeSplitting) {
-  config.experimentalCodeSplitting = true;
-  config.experimentalDynamicImport = true;
```
Comment (gregmagolan) [inline internal/rollup/rollup.config.js]: Nice. Looks like the feature is no longer experimental since I worked on this 👍

File: internal/e2e/rollup_code_splitting/BUILD.bazel [hunk 6]
```diff
@@ -7,7 +7,19 @@ rollup_bundle(
...
+)
+
+rollup_bundle(
+    name = "bundle_multi_entry",
+    srcs = glob(
+        ["*.js"],
+        exclude = ["*.spec.js"],
+    ),
     additional_entry_points = ["build_bazel_rules_nodejs/internal/e2e/rollup_code_splitting/additional_entry.js"],
+    enable_code_splitting = False,
```
Comment (gregmagolan) [inline internal/e2e/rollup_code_splitting/BUILD.bazel]: This would implied to be true because `additional_entry_points` is specified?

Comment (alexeagle) [reply internal/e2e/rollup_code_splitting/BUILD.bazel]: yeah this should have been an error, will fix

**Comment by alexeagle** [comment]
I'll try it in angular-bazel-example before next release - want to get Jesse's PR in under his authorship before I do more with it
